content/blog/2022/04/securing-kubewarden-policies.md

```diff
@@ -0,0 +1,209 @@
+are with an OpenID Connect service (SSO via your own Okta instance, Github,
+Google, etc). Nobody (neither developers, project leads, sponsors...) needs to
+have access to keys (hence "keyless") and Sigstore never obtains your private
+key.
+
+Once the artifact is signed, the proof of signature is then send to an
+append-only transparency log ([Rekor](https://github.com/sigstore/rekor)) that
+allows monitoring of such signatures and protects against timing attacks.
+
+Since there's no need for key secrets and the like in Keyless mode, it is easily
+automated inside CIs and implemented and monitored in the open. Hence why it is
+so interesting.
+
+### Building a Rust Sigstore stack
+
+The part of the Kubewarden stack that instantiates and runs policies (the
+[policy server](https://github.com/kubewarden/policy-server) and libs) is
+written in Rust. This meant we needed a good Rust implementation of Sigstore
+features, and one didn't exist. We are glad to say we have created a new crate,
+[sigstore-rs](https://github.com/sigstore/sigstore-rs), under the Sigstore org
+in an upstream-first manner, and it is taking a life of its own.
+
+## Securing kubewarden policies
+
+As you may already know, Kubewarden Policies are just small wasm-compiled
+executables (~1 to ~6MBs) that they are shipped as OCI images. Let see how
+Kubewarden protects policies against Secure Supply Chain attacks by signing and
+verifying them before they run.
+
+### Signing your Kubewarden Policy
+
+Signing a Policy, or what is the same, a container image, means just adding a
+new layer with the signature. In the Sigstore workflow, one can sign with
+Public-Private keypair, or Keyless. Both can also add `key=value` annotations to
+the signatures.
+
+The Public-Private key pair signing is straightforward, using [sigstore/cosign](https://github.com/SigStore/cosign):
+
+```console
+$ COSIGN_PASSWORD=yourpass cosign generate-key-pair
+
+Private key written to cosign.key
+Public key written to cosign.pub
+
+$ COSIGN_PASSWORD=yourpass cosign sign \
+  --key cosign.key --annotations blog=yes \
+  ghcr.io/kubewarden/policies/user-group-psp:v0.2.0
+
+Pushing signature to: ghcr.io/kubewarden/policies/user-group-psp
+```
+
+The Keyless mode is more interesting:
+```console
+$ COSIGN_EXPERIMENTAL=1 cosign sign \
+  --annotations blog=yes \
+  ghcr.io/kubewarden/policies/user-group-psp:v0.2.0
+
+Generating ephemeral keys...
+Retrieving signed certificate...
+Your browser will now be opened to:
+https://oauth2.sigstore.dev/auth/auth?access_type=online&client_id=sigstore&code_challenge=(...)
+Successfully verified SCT...
+tlog entry created with index: (...)
+Pushing signature to: ghcr.io/viccuad/policies/volumes-psp
+```
+
+What happened? `cosign` prompted us for an OpenID Connect provider on the
+browser, which authenticated us, and allowed Fulcio to sign our ephemeral keys.
+
+If this were to happen in a CI, for example GH Actions, the CI would provide an authentication token.
```

Comment:
we can mention that this works in our GH action because cosign has support for detecting some automated environments and producing an identity token. It currently supports github and google. If they want to use a different CI users can provide an `--indentity-token`. more info https://github.com/sigstore/cosign/blob/main/KEYLESS.md#identity-tokens

Reply:
That sounds better, yes. Done in https://github.com/kubewarden/kubewarden.io/pull/82/commits/ce239ddb36369430df124377bc535a8f4d9ecf42.